src/server/db_slice.cc

```diff
@@ -1135,6 +1134,17 @@ DbSlice::PrimeItAndExp DbSlice::ExpireIfNeeded(const Context& cntx, PrimeIterato
   return {PrimeIterator{}, ExpireIterator{}};
 }
 
+// It's fine to preempt if we hold a lock to the key because there won't be any
+// concurrent changes. If preempt = false, we need to Wait on the blocking counter.
+DbSlice::PrimeItAndExp DbSlice::ExpireIfNeeded(const Context& cntx, PrimeIterator it,
```

Comment:
Instead of this funciton I would pass the preemptive param to ExpireIfNeededImpl and propagate to RecordExpiry so that we will we will be able to call flush to sync on the preemptive flows
this way you can put the fiber gaurd in the record exipry if we have preemptive=flase